cub/cub/device/dispatch/dispatch_streaming_reduce.cuh

```diff
@@ -0,0 +1,451 @@
+  _CCCL_HOST_DEVICE _CCCL_FORCEINLINE KeyValuePair<GlobalOffsetT, AccumT>
+  operator()(KeyValuePair<PerPartitionOffsetT, AccumT> partition_aggregate)
+  {
+    return KeyValuePair<GlobalOffsetT, AccumT>{
+      current_partition_offset + static_cast<GlobalOffsetT>(partition_aggregate.key), partition_aggregate.value};
+  }
+};
+
+/**
+ * Offsets a the given input iterator by a fixed offset, such that when an item at index `i` is accessed, the item
+ * `it[*offset_it + i]` is accessed.
+ */
+template <typename Iterator, typename OffsetItT>
+class offset_iterator : public THRUST_NS_QUALIFIER::iterator_adaptor<offset_iterator<Iterator, OffsetItT>, Iterator>
+{
+public:
+  using super_t = THRUST_NS_QUALIFIER::iterator_adaptor<offset_iterator<Iterator, OffsetItT>, Iterator>;
+
+  _CCCL_HOST_DEVICE _CCCL_FORCEINLINE offset_iterator(const Iterator& it, OffsetItT offset_it)
+      : super_t(it)
+      , offset_it(offset_it)
+  {}
+
+  // befriend thrust::iterator_core_access to allow it access to the private interface below
+  friend class THRUST_NS_QUALIFIER::iterator_core_access;
+
+private:
+  OffsetItT offset_it;
+
+  _CCCL_HOST_DEVICE _CCCL_FORCEINLINE typename super_t::reference dereference() const
+  {
+    return *(this->base() + (*offset_it));
+  }
+};
+
+template <typename Iterator, typename OffsetItT>
+_CCCL_HOST_DEVICE offset_iterator<Iterator, OffsetItT> make_offset_iterator(Iterator it, OffsetItT offset_it)
+{
+  return offset_iterator<Iterator, OffsetItT>{it, offset_it};
+}
+
+/**
+ * Helper transform operation that writes to the (index, extremum)-key-value-pair to the user-provided output iterator.
+ * This helper allows "implicit conversion" between KeyValuePair types that have a different key type, which may happen
+ * if the user uses a different index type than the global offset type used by the algorithm.
+ */
+template <typename KeyValuePairOutItT>
+struct write_to_user_out_it
+{
+  KeyValuePairOutItT out_it;
+
+  template <typename IndexT, typename OffsetT, typename ResultT>
+  _CCCL_DEVICE _CCCL_FORCEINLINE void operator()(IndexT, cub::KeyValuePair<OffsetT, ResultT> reduced_result)
+  {
+    using kv_pair_t = ::cuda::std::_If<
+      ::cuda::std::is_assignable<decltype(*out_it), cub::KeyValuePair<::cuda::std::uint64_t, ResultT>>::value,
+      cub::KeyValuePair<::cuda::std::uint64_t, ResultT>,
+      ::cuda::std::_If<
+        ::cuda::std::is_assignable<decltype(*out_it), cub::KeyValuePair<::cuda::std::int64_t, ResultT>>::value,
+        cub::KeyValuePair<::cuda::std::int64_t, ResultT>,
+        ::cuda::std::_If<
+          ::cuda::std::is_assignable<decltype(*out_it), cub::KeyValuePair<::cuda::std::uint32_t, ResultT>>::value,
+          cub::KeyValuePair<::cuda::std::uint32_t, ResultT>,
+          cub::KeyValuePair<::cuda::std::int32_t, ResultT>>>>;
+
+    using index_t = typename kv_pair_t::Key;
+    *out_it       = kv_pair_t{static_cast<index_t>(reduced_result.key), reduced_result.value};
```

Comment:
Important: this is buggy. If `index_t` is too small to hold the resulting index the user gets a wrong result. I explained a different API design at length above. If you don't want to change it, I would at least want to have an assertion here that the narrowing cast does not change the value.

Btw, this may be a use case for `gsl::narrow` #2278

Reply:
Important: this comment is marked as resolved, but I don't see an assert that checks whether `static_cast<index_t>(reduced_result.key)` does change the value. You could add:
```suggestion
    _CCCL_ASSERT(static_cast<OffsetT>(static_cast<index_t>(reduced_result.key)) == reduced_result.key);
    *out_it       = kv_pair_t{static_cast<index_t>(reduced_result.key), reduced_result.value};
```

Reply:
Sorry, I probably missed that.